### PR TITLE
Automatic Desired Area Function for XYDelaunay

### LINKED ISSUE #27433: Adding Automatically Generated Desired Area Function for XYDelaunayGenerator
### Motivation

Current XYDelaunayGenerator (and the MGs using XYDelaunayGenerator as their subgenerator) takes either a uniform element area limiting function or a user defined parsed area function. It is not straightforward for users to provide such a function due to the complexity of the meshing geometry.

It would be convenient to automatically generate an area function to adapt the input external and holes boundaries.

### Design

Based on the corresponding updates in libMesh https://github.com/libMesh/libmesh/pull/3813, enable the capability in MOOSE.

### Impact

Enhance user experience by providing an automatic option to improve the mesh quality for XYDelaunayGenerator

### PR DESCRIPTION
**Pull request by miaoyinb**
closes #27433

Pending https://github.com/libMesh/libmesh/pull/3813

### CODE REVIEW

**Comment by miaoyinb** [comment]
@GiudGiud @roystgnr 
I was wondering what is the routine to do a PR that needs a libMesh update. Any suggestion would be appreciated.

Meanwhile, I will try to fix the current error, I think it is related to the randomness in triangulation when using different numbers of cores.

**Comment by GiudGiud** [comment]
Usually we do libmesh updates in a separate PR that updates the submodule. There's nothing too special about it except you need to tell Logan or Jason about it so they look at / add deployment items in the PR

**Comment by miaoyinb** [comment]
@GiudGiud I think all the tests are good except for those relying on conda. 

@milljm @loganharbour I would appreciate it a lot if you could help with the libmesh submodule update. Thanks in advance.

File: framework/doc/content/source/meshgenerators/XYDelaunayGenerator.md
Comment (GiudGiud) [inline framework/doc/content/source/meshgenerators/XYDelaunayGenerator.md:12]:
```suggestion
be meshed by the mesh generators specified in the [!param](//Mesh/XYDelaunayGenerator/holes) vector parameter.
```

File: framework/src/meshgenerators/XYDelaunayGenerator.C
Comment (GiudGiud) [inline framework/src/meshgenerators/XYDelaunayGenerator.C:80]:
```suggestion
  params.addParam<bool>("use_auto_area_func", false, "Use the automatic area function for the triangle meshing region.");
```

File: framework/src/meshgenerators/XYDelaunayGenerator.C
Comment (GiudGiud) [inline framework/src/meshgenerators/XYDelaunayGenerator.C:111]:
```suggestion
      "Automatic triangle meshing area control");
```
I would add the other parameters for that region in another parameter group

File: modules/reactor/src/meshgenerators/FlexiblePatternGenerator.C
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/FlexiblePatternGenerator.C:88]:
```suggestion
  params.addParam<bool>("use_auto_area_func", false, "Use the automatic area function for triangle-meshing in the background.");
```

File: modules/reactor/src/meshgenerators/PeripheralTriangleMeshGenerator.C
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/PeripheralTriangleMeshGenerator.C:71]:
```suggestion
                              "Peripheral Area Delaunay");
```

File: modules/reactor/src/meshgenerators/PeripheralTriangleMeshGenerator.C
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/PeripheralTriangleMeshGenerator.C:41]:
```suggestion
  params.addParam<bool>("use_auto_area_func", false, "Use the automatic area function in the peripheral region.");
```

File: test/tests/meshgenerators/xy_delaunay_generator/tests
Comment (GiudGiud) [inline test/tests/meshgenerators/xy_delaunay_generator/tests:132]:
```suggestion
      detail = "if more than one element area limiting methods are specified by the user"
```

File: modules/reactor/doc/content/source/meshgenerators/PeripheralTriangleMeshGenerator.md
Comment (GiudGiud) [inline modules/reactor/doc/content/source/meshgenerators/PeripheralTriangleMeshGenerator.md:9]:
```suggestion
Optionally, triangle refinement parameters can be used to improve the quality of the triangles in the periphery region. These are specified with either the [!param](/Mesh/PeripheralTriangleMeshGenerator/desired_area) parameter, which specifies a single desired maximum triangle area, or the [!param](/Mesh/PeripheralTriangleMeshGenerator/desired_area_func) parameter, which specifies the desired maximum triangle area as a function of the spatial coordinates x and y. Alternatively, this type of refinement can be done following an automatically generated area function. This feature is enabled by [!param](/Mesh/XYDelaunayGenerator/use_auto_area_func), while the related parameters include [!param](/Mesh/XYDelaunayGenerator/auto_area_func_default_size), [!param](/Mesh/XYDelaunayGenerator/auto_area_func_default_size_dist), [!param](/Mesh/XYDelaunayGenerator/auto_area_function_num_points), and [!param](/Mesh/XYDelaunayGenerator/auto_area_function_power). See the full documentation for [`XYDelaunayGenerator`](XYDelaunayGenerator.md) for more details on the area refinement options.
```

**Review by GiudGiud — COMMENTED** [review]
looks good!
CIVET's message tells you which files to modify if you want to try to get these tests green. We will have to have the deployment team review it regardless though so feel free to leave that to them if you prefer

File: framework/doc/content/source/meshgenerators/XYDelaunayGenerator.md
Comment (GiudGiud) [inline framework/doc/content/source/meshgenerators/XYDelaunayGenerator.md:60]:
```suggestion
using one of the following three approaches:
```

File: framework/src/meshgenerators/XYDelaunayGenerator.C
Comment (GiudGiud) [inline framework/src/meshgenerators/XYDelaunayGenerator.C:141]: We ll want to check for all the automatic area parameters here instead. Otherwise they can be specified by the user then ignored because another method is in use

**Comment by miaoyinb** [comment]
> looks good! CIVET's message tells you which files to modify if you want to try to get these tests green. We will have to have the deployment team review it regardless though so feel free to leave that to them if you prefer

I will try it just out of curiosity.

**Comment by milljm** [comment]
@miaoyinb well done on the Conda side of things! There exists one more thing required: updating the `versioner_hashes.yaml` file in `moose/scripts/tests`.

It is generally the very last thing you want to do. And because it looks like some things are still failing, you will want to hold off on the following instructions until then.

The version hash file allows us to keep a solid record tree on all the updates occurring with our Conda packages (as well as Apptainer definition files).

> [!TIP]
> The following instructions are also available within and at the beginning of the following file: `moose/scripts/tests/versioner_hashes.yaml`


1. get the current HASH of your completed PR:
    ```bash
    git rev-parse HEAD
    ```
    record this information for later use.
2. open the file `moose/scripts/tests/versioner_hashes.yaml` and append the above output while also providing the MOOSE PR responsible for making these changes using the following syntax:
    ```pre
    <HASH>: #27433
    ```
3. run our `versioner.py` script located in `moose/scripts` on all libraries we keep tabs on (libMesh being one of them, thus why you need to do these steps):
    ```bash
    ./versioner.py mpich
    ./versioner.py petsc
    ./versioner.py libmesh
    ./versioner.py moose-dev
    ./versioner.py wasp
    ```
5. add each key and assign the corresponding value gathered from the above output, using the following syntax:
    ```pre
    <HASH>: #27433
      mpich: <hash>
      petsc: <hash>
      libmesh: <hash>
      moose-dev: <hash>
      wasp: <hash>
    ```
    *because you are only updating libMesh, you should see no changes for mpich, petsc, and wasp entries from the previous HASH: #PR entry*
6. `git add; git commit; git push` one last time.

**Comment by miaoyinb** [comment]
I do not think the "controlled app tests" failure is related, but I cannot see or invalidate it.

**Comment by GiudGiud** [comment]
@milljm  can you please sign off on the libmesh update items?
this is otherwise ready to merge

**Comment by miaoyinb** [comment]
@milljm @GiudGiud Let me know if any additional modifications are needed. Thanks.

**Comment by GiudGiud** [comment]
@miaoyinb looks like somebody updated libmesh elsewhere?

**Comment by GiudGiud** [comment]
Squashed minor commits and removed libmesh update. Will merge if libmesh submodule is far enough forward now

**Comment by miaoyinb** [comment]
@GiudGiud I am on travel this week and just landed. Will work on this. Thanks.

**Comment by miaoyinb** [comment]
I think `moose-dev` and `wasp` have been updated, but `moose-libmesh` seems not...

**Comment by miaoyinb** [comment]
@GiudGiud it seems you already worked on it. Thanks. Do I need to do anything else?

**Comment by GiudGiud** [comment]
should be fine once I get those details right

**Comment by milljm** [comment]
The Conda stuff looks good to me! Well Done.

I need to allow others that know more about libMesh to approve the PR.

**Review by GiudGiud — APPROVED** [review]
@roystgnr signed off on updating in a pm

**Comment by GiudGiud** [comment]
Thanks for the contribution!
I think moving forward we'll want to keep the libmesh updates in separate PRs. It will be easier to bring all the relevant expertise together

**Comment by loganharbour** [comment]
This broke next https://civet.inl.gov/event/180188/

**Comment by miaoyinb** [comment]
7500cf899b0b838167fae6f5a1eb5c7a63dc2f09

This is still caused by randomness in triangulation. I did a comprehensive test and this commit should fix it.
What's the best way to fix this? A new PR?

**Comment by GiudGiud** [comment]
new PR is the way

**Pull request merged**
